Sources/SwiftBSON/ExtendedJSONDecoder.swift

```diff
@@ -13,6 +15,32 @@ public class ExtendedJSONDecoder {
         return formatter
     }()
 
+    /// A set of all the possible extenededJSON wrapper keys.
+    private static var wrapperKeySet: Set<String> = {
+        Set(ExtendedJSONDecoder.wrapperKeyMap.keys)
+    }()
+
+    /// A map from extended JSON wrapper keys (e.g. "$numberLong") to the BSON type(s) that they correspond to.
+    ///
+    /// Some types are associated with multiple wrapper keys (e.g. "$code" and "$scope" both map to
+    /// `BSONCodeWithScope`). Some wrapper keys are associated with multiple types (e.g. "$code" maps to both
+    /// `BSONCode` and `BSONCodeWithScope`). Attempt to decode each of the types returned from the map until one works
+    /// to find the proper decoding.
+    private static var wrapperKeyMap: [String: [BSONValue.Type]] = {
```

Comment:
this allows us to use the input JSON to determine which type(s) to attempt to decode rather than just trying each in succession. Code and CodeWithScope make this a little trickier, hence it not being a 1:1 mapping.